contracts/strategies/donation-voring-merkle-distribution/DonationVotingMerkleDistributionStrategy.sol

```diff
@@ -0,0 +1,602 @@
+    /// @param _allocationStartTime The start time for the allocation
+    /// @param _allocationEndTime The end time for the allocation
+    function updatePoolTimestamps(
+        uint256 _registrationStartTime,
+        uint256 _registrationEndTime,
+        uint256 _allocationStartTime,
+        uint256 _allocationEndTime
+    ) external onlyPoolManager(msg.sender) {
+        _isPoolTimestampValid(_registrationStartTime, _registrationEndTime, _allocationStartTime, _allocationEndTime);
+
+        registrationStartTime = _registrationStartTime;
+        registrationEndTime = _registrationEndTime;
+        allocationStartTime = _allocationStartTime;
+        allocationEndTime = _allocationEndTime;
+
+        emit TimestampsUpdated(
+            registrationStartTime, registrationEndTime, allocationStartTime, allocationEndTime, msg.sender
+        );
+    }
+
+    /// @notice Withdraw funds from pool
+    /// @param _amount The amount to be withdrawn
+    function withdraw(uint256 _amount) external onlyPoolManager(msg.sender) onlyAfterAllocation {
+        if (block.timestamp <= allocationEndTime + 30 days) {
+            revert NOT_ALLOWED();
+        }
+
+        // TODO: discuss if we want to allow withdrawing ONLY "poolAmount - totalPayoutAmount" from the pool
+        IAllo.Pool memory pool = allo.getPool(poolId);
+        if (poolAmount - totalPayoutAmount < _amount) {
+            revert NOT_ALLOWED();
+        }
+
+        poolAmount -= _amount;
+        _transferAmount(pool.token, msg.sender, _amount);
+    }
+
+    /// @notice Claim allocated tokens
+    /// @param _claims Claims to be claimed
+    function claim(Claim[] calldata _claims) external nonReentrant onlyAfterAllocation {
+        uint256 claimsLength = _claims.length;
+        for (uint256 i; i < claimsLength;) {
+            Claim memory singleClaim = _claims[i];
+            Recipient memory recipient = _recipients[singleClaim.recipientId];
+            uint256 amount = claims[singleClaim.recipientId][singleClaim.token];
+
+            if (amount == 0) {
+                revert INVALID();
+            }
+
+            claims[singleClaim.recipientId][singleClaim.token] = 0;
+
+            address token = singleClaim.token;
+
+            _transferAmount(token, recipient.recipientAddress, amount);
+
+            emit Claimed(singleClaim.recipientId, recipient.recipientAddress, amount, token);
+            unchecked {
+                i++;
+            }
+        }
+    }
+
+    /// ==================================
+    /// ============ Merkle ==============
+    /// ==================================
+
+    /// @notice Invoked by round operator to update the merkle root and distribution MetaPtr
+    /// @param encodedDistribution encoded distribution
+    function updateDistribution(bytes calldata encodedDistribution)
+        external
+        onlyAfterAllocation
+        onlyPoolManager(msg.sender)
+    {
+        if (merkleRoot != "") {
```

Comment:
This means merkleroot can be set only once.
Not sure if this is intended behaviour or we want to allow resetting distribution multiple times